File: javaclient/src/main/java/jagex2/dash3d/AnimFrame.java
package jagex2.dash3d;

import deob.ObfuscatedName;
import jagex2.io.Packet;

@ObfuscatedName("h")
public class AnimFrame {

	@ObfuscatedName("h.c")
	public static AnimFrame[] instances;

	@ObfuscatedName("h.d")
	public int delay;

	@ObfuscatedName("h.e")
	public AnimBase base;

	@ObfuscatedName("h.f")
	public int size;

	@ObfuscatedName("h.g")
	public int[] ti; // transform indices (groups affected by this frame)

	@ObfuscatedName("h.h")
	public int[] tx; // transform x

	@ObfuscatedName("h.i")
	public int[] ty; // transform y

	@ObfuscatedName("h.j")
	public int[] tz; // transform z

	@ObfuscatedName("h.a(I)V")
	public static void init(int count) {
		instances = new AnimFrame[count + 1];
	}

	@ObfuscatedName("h.a(Z[B)V")
	public static void unpack(byte[] src) {
		Packet buf = new Packet(src);

		buf.pos = src.length - 8;
		int headLen = buf.g2();
		int tran1Len = buf.g2();
		int tran2Len = buf.g2();
		int delLen = buf.g2();
		int pos = 0;

		Packet head = new Packet(src);
		head.pos = pos;
		pos = pos + headLen + 2;

		Packet tran1 = new Packet(src);
		tran1.pos = pos;
		pos = pos + tran1Len;

		Packet tran2 = new Packet(src);
		tran2.pos = pos;
		pos = pos + tran2Len;

		Packet del = new Packet(src);
		del.pos = pos;
		pos = pos + delLen;

		Packet var16 = new Packet(src);
		var16.pos = pos;

		AnimBase base = new AnimBase(var16);

		int total = head.g2();
		int[] tempTi = new int[500];
		int[] tempTx = new int[500];
		int[] tempTy = new int[500];
		int[] tempTz = new int[500];
		for (int i = 0; i < total; i++) {
			int id = head.g2();

			AnimFrame frame = instances[id] = new AnimFrame();
			frame.delay = del.g1();
			frame.base = base;

			int groupCount = head.g1();
			int lastGroup = -1;
			int current = 0;
			for (int j = 0; j < groupCount; j++) {
				int flags = tran1.g1();
				if (flags > 0) {
					if (base.types[j] != 0) {
						for (int group = j - 1; group > lastGroup; group--) {
							if (base.types[group] == 0) {
								tempTi[current] = group;
								tempTx[current] = 0;
								tempTy[current] = 0;
								tempTz[current] = 0;
								current++;
								break;
							}
						}
					}

					tempTi[current] = j;

					short defaultValue = 0;
					if (base.types[tempTi[current]] == 3) {
						defaultValue = 128;
					}

					if ((flags & 0x1) == 0) {
						tempTx[current] = defaultValue;
					} else {
						tempTx[current] = tran2.gsmart();
					}

					if ((flags & 0x2) == 0) {
						tempTy[current] = defaultValue;
					} else {
						tempTy[current] = tran2.gsmart();
					}

					if ((flags & 0x4) == 0) {
						tempTz[current] = defaultValue;
					} else {
						tempTz[current] = tran2.gsmart();
					}

					lastGroup = j;
					current++;
				}
			}

			frame.size = current;
			frame.ti = new int[current];
			frame.tx = new int[current];
			frame.ty = new int[current];
			frame.tz = new int[current];

			for (int j = 0; j < current; j++) {
				frame.ti[j] = tempTi[j];
				frame.tx[j] = tempTx[j];
				frame.ty[j] = tempTy[j];
				frame.tz[j] = tempTz[j];
			}
		}
	}

	@ObfuscatedName("h.a(B)V")
	public static void unload() {
		instances = null;
	}

	@ObfuscatedName("h.a(II)Lh;")
	public static AnimFrame get(int id) {
		if (instances == null) {
			return null;
		} else {
			return instances[id];
		}
	}
}

File: engine/src/cache/graphics/AnimFrame.ts
import AnimBase from '#/cache/graphics/AnimBase.js';
import OnDemand from '#/engine/OnDemand.js';
import Packet from '#/io/Packet.js';

export default class AnimFrame {
    static instances: AnimFrame[] = [];
    static order: number[] = [];

    delay: number = 0;
    base: number = 0;
    length: number = 0;
    groups: Int32Array = new Int32Array();
    x: Int32Array = new Int32Array();
    y: Int32Array = new Int32Array();
    z: Int32Array = new Int32Array();

    static load() {
        const count = OnDemand.cache.count(2);
        for (let i = 0; i < count; i++) {
            const data = OnDemand.cache.read(2, i, true);
            if (data) {
                AnimFrame.unpack(data);
            }
        }
    }

    static unpack(src: Uint8Array) {
        const meta = new Packet(src);
        meta.pos = src.length - 8;

        let offset = 0;
        const head = new Packet(src);
        head.pos = offset;
        offset += meta.g2() + 2;

        const tran1 = new Packet(src);
        tran1.pos = offset;
        offset += meta.g2();

        const tran2 = new Packet(src);
        tran2.pos = offset;
        offset += meta.g2();

        const del = new Packet(src);
        del.pos = offset;
        offset += meta.g2();

        const baseData = new Packet(src);
        baseData.pos = offset;
        const baseId = AnimBase.unpack(baseData);

        const total = head.g2();
        const bases = new Int32Array(500);
        const x = new Int32Array(500);
        const y = new Int32Array(500);
        const z = new Int32Array(500);

        for (let i = 0; i < total; i++) {
            const id = head.g2();
            AnimFrame.order.push(id);

            const frame = new AnimFrame();
            frame.delay = del.g1();
            frame.base = baseId;

            const groupCount = head.g1();
            let lastGroup = -1;
            let length = 0;

            for (let group = 0; group < groupCount; group++) {
                const flags = tran1.g1();
                if (flags === 0) {
                    continue;
                }

                if (AnimBase.instances[baseId].types[group] !== AnimBase.OP_BASE) {
                    for (let cur = group - 1; cur > lastGroup; cur--) {
                        if (AnimBase.instances[baseId].types[cur] === AnimBase.OP_BASE) {
                            bases[length] = cur;
                            x[length] = 0;
                            y[length] = 0;
                            z[length] = 0;
                            length++;
                            break;
                        }
                    }
                }

                bases[length] = group;

                let defaultValue = 0;
                if (AnimBase.instances[baseId].types[group] === AnimBase.OP_SCALE) {
                    defaultValue = 128;
                }

                if ((flags & 0x1) != 0) {
                    x[length] = tran2.gsmart();
                } else {
                    x[length] = defaultValue;
                }

                if ((flags & 0x2) != 0) {
                    y[length] = tran2.gsmart();
                } else {
                    y[length] = defaultValue;
                }

                if ((flags & 0x4) != 0) {
                    z[length] = tran2.gsmart();
                } else {
                    z[length] = defaultValue;
                }

                lastGroup = group;
                length++;
            }

            frame.length = length;
            frame.groups = new Int32Array(length);
            frame.x = new Int32Array(length);
            frame.y = new Int32Array(length);
            frame.z = new Int32Array(length);

            for (let j = 0; j < length; j++) {
                frame.groups[j] = bases[j];
                frame.x[j] = x[j];
                frame.y[j] = y[j];
                frame.z[j] = z[j];
            }

            AnimFrame.instances[id] = frame;
        }
    }
}
